### PR TITLE
add recurring filter

### LINKED ISSUE #179: sort recurring expenses by category

### PR DESCRIPTION
**Pull request by bexrr**
closes #179

### CODE REVIEW

**Comment by bexrr** [comment]
@Agreon please review

File: server/recurring.go [hunk 1]
```diff
@@ -100,17 +121,81 @@ func listRecurring(c *gin.Context) {
...
 		return
 	}
 
-	err = db.Get(&recurring.Entries, "SELECT count(*) FROM recurring WHERE user_id=$1 AND parent_id IS NULL AND is_expense=$2", userID, isExpense)
-	if err != nil {
-		saveErrorInfo(c, err, 500)
-		return
+	page := c.MustGet("page")
+	endDate, existingEndDate := filterOptions["end_date<"]
+	startDate, existingStartDate := filterOptions["start_date>"]
```
Comment (Agreon) [inline server/recurring.go:133]:
```suggestion
	endDate, endDateExists := filterOptions["end_date<"]
	startDate, startDateExists := filterOptions["start_date>"]
```

File: server/recurring.go [hunk 2]
```diff
@@ -100,17 +121,81 @@ func listRecurring(c *gin.Context) {
...
+				relevant_recurring.category,
+				relevant_recurring.is_expense,
+				relevant_recurring.costs AS costs,
+				relevant_recurring.start_date_return_value AS start_date,
+				relevant_recurring.end_date_return_value AS end_date,
+				relevant_recurring.created_at,
+				relevant_recurring.updated_at
+			FROM (
+				SELECT
+					id,	name, costs, user_id, category, is_expense,
```
Comment (Agreon) [inline server/recurring.go:152]:
```suggestion
					id, name, costs, user_id, category, is_expense, 
```

File: server/recurring.go [hunk 3]
```diff
@@ -100,17 +121,81 @@ func listRecurring(c *gin.Context) {
...
+		}
+		err = dbExtended.GetWithFilterOptions(&recurring.Entries, `
+			SELECT count(*)  
+			FROM (
+				SELECT
+					id
+				FROM recurring
+				WHERE
+					user_id=$1 AND is_expense=$2 AND start_date<$3 AND (end_date>$4 OR end_date=$5)
+			) AS relevant_recurring
```
Comment (Agreon) [inline server/recurring.go:181]:
```suggestion
			SELECT count(*)  
				FROM recurring
				WHERE
					user_id=$1 AND is_expense=$2 AND start_date<$3 AND (end_date>$4 OR end_date=$5)
			
```

**Review by Agreon — CHANGES_REQUESTED** [review]
Good work, almost :hankey:

**Review by Agreon — APPROVED** [review]

**Pull request merged**